runtime/src/accounts_db.rs

```diff
@@ -1883,35 +1893,48 @@ impl AccountsDB {
     }
 
     /// Scan a specific slot through all the account storage in parallel
-    pub fn scan_account_storage<F, B>(&self, slot: Slot, scan_func: F) -> Vec<B>
-    where
-        F: Fn(LoadedAccount, AppendVecId, &mut B) + Send + Sync,
-        B: Send + Default,
-    {
-        self.scan_account_storage_inner(slot, scan_func)
-    }
-
-    fn scan_account_storage_inner<F, B>(&self, slot: Slot, scan_func: F) -> Vec<B>
+    pub fn scan_account_storage<C, F, R, B>(
+        &self,
+        slot: Slot,
+        cache_map_func: C,
+        storage_scan_func: F,
+    ) -> ScanStorageResult<R, B>
     where
-        F: Fn(LoadedAccount, AppendVecId, &mut B) + Send + Sync,
-        B: Send + Default,
+        C: Fn(LoadedAccount) -> Option<R> + Sync,
+        R: Send,
+        F: Fn(&B, LoadedAccount) + Sync,
+        B: Send + Default + Sync,
     {
         if let Some(slot_cache) = self.accounts_cache.slot_cache(slot) {
             // If we see the slot in the cache, then all the account information
             // is in this cached slot
-            let mut retval = B::default();
-            for cached_account in slot_cache.iter() {
-                scan_func(
-                    LoadedAccount::Cached((
-                        *cached_account.key(),
-                        Cow::Borrowed(cached_account.value()),
-                    )),
-                    CACHE_VIRTUAL_STORAGE_ID,
-                    &mut retval,
-                );
+            if slot_cache.len() > SCAN_SLOT_PAR_ITER_THRESHOLD {
```

Comment:
well, why can't we use `.chunks()`?

Reply:
hmmm this would affect thread work distribution right? If we have`SCAN_SLOT_PAR_ITER_THRESHOLD + 1` items, one thread would be doing work on `SCAN_SLOT_PAR_ITER_THRESHOLD` items, while the other on only `1` item